internal/terraform/context_plan_ephemeral_test.go

```diff
@@ -7,135 +7,534 @@ import (
+        }
+    }
+}
+ephemeral "ephem_resource" "data" {}
+
+# We expect this to error since it should be an ephemeral value
+output "value" {
+    value = provider::ephem::either(ephemeral.ephem_resource.data.value, "b")
 }
 `,
-	})
+				"main.tf": `
+module "child" {
+    source = "./child"
+}
+				`,
+			},
+
+			expectValidateDiagnostics: func(m *configs.Config) (diags tfdiags.Diagnostics) {
+				return diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Ephemeral value not allowed",
+					Detail:   "This output value is not declared as returning an ephemeral value, so it cannot be set to a result derived from an ephemeral value.",
+				})
+			},
+		},
+
+		"check blocks": {
+			toBeImplemented: true,
+			module: map[string]string{
+				"main.tf": `
+ephemeral "ephem_resource" "data" {}
+
+check "check_using_ephemeral_value" {
+  assert {
+    condition = ephemeral.ephem_resource.data.bool
+    error_message = "This should not fail"
+  }
+}
+				`,
+			},
+			expectOpenEphemeralResourceCalled:           true,
+			expectValidateEphemeralResourceConfigCalled: true,
+			expectCloseEphemeralResourceCalled:          true,
+
+			assertPlan: func(t *testing.T, p *plans.Plan) {
+				// Checks using ephemeral values should not be included in the plan
+				if p.Checks.ConfigResults.Len() > 0 {
+					t.Fatalf("Expected no checks to be included in the plan, but got %d", p.Checks.ConfigResults.Len())
+				}
+			},
+		},
 
-	ephem := &testing_provider.MockProvider{
-		GetProviderSchemaResponse: &providers.GetProviderSchemaResponse{
-			EphemeralResourceTypes: map[string]providers.Schema{
-				"ephem_resource": {
-					Block: &configschema.Block{
-						Attributes: map[string]*configschema.Attribute{
-							"value": {
-								Type:     cty.String,
-								Computed: true,
+		"function ephemeralasnull": {
+			module: map[string]string{
+				"child/main.tf": `
+ephemeral "ephem_resource" "data" {}
+
+output "value" {
+    value = ephemeralasnull(ephemeral.ephem_resource.data.value)
+}
+`,
+				"main.tf": `
+module "child" {
+    source = "./child"
+}
+				`,
+			},
+			expectOpenEphemeralResourceCalled:           true,
+			expectValidateEphemeralResourceConfigCalled: true,
+			expectCloseEphemeralResourceCalled:          true,
+		},
+
+		"function ephemeral": {
+			toBeImplemented: true,
+			module: map[string]string{
+				"child/main.tf": `
+
+# We expect this to error since it should be an ephemeral value
+output "value" {
+    value = ephemeral("hello world")
+}
+`,
+				"main.tf": `
+module "child" {
+    source = "./child"
+}
+			`,
+			},
+
+			expectValidateDiagnostics: func(m *configs.Config) (diags tfdiags.Diagnostics) {
+				return diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Ephemeral value not allowed",
+					Detail:   "This output value is not declared as returning an ephemeral value, so it cannot be set to a result derived from an ephemeral value.",
+				})
+			},
+		},
```

Comment:
```suggestion
```

Just to avoid possible dead code that will be harder to track, shall we maybe link to https://github.com/hashicorp/terraform/pull/35653 and/or add the test there once this is merged?